File: apps/api/plane/ee/urls/app/workflow.py
# SPDX-FileCopyrightText: 2023-present Plane Software, Inc.
# SPDX-License-Identifier: LicenseRef-Plane-Commercial
#
# Licensed under the Plane Commercial License (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# https://plane.so/legals/eula
#
# DO NOT remove or modify this notice.
# NOTICE: Proprietary and confidential. Unauthorized use or distribution is prohibited.

# Django imports
from django.urls import path

# Module imports
from plane.ee.views.app.workflow import (
    WorkflowEndpoint,
    WorkflowActivityEndpoint,
    WorkflowTransitionEndpoint,
    WorkflowTransitionApproverEndpoint,
)


urlpatterns = [
    path(
        "workspaces/<str:slug>/workflow-states/",
        WorkflowEndpoint.as_view(),
        name="project-workflows",
    ),
    path(
        "workspaces/<str:slug>/workflow-states/<uuid:state_id>/",
        WorkflowEndpoint.as_view(),
        name="project-workflows",
    ),
    path(
        "workspaces/<str:slug>/projects/<uuid:project_id>/workflow-reset/",
        WorkflowEndpoint.as_view(),
        name="project-workflows",
    ),
    path(
        "workspaces/<str:slug>/projects/<uuid:project_id>/workflow-transitions/",
        WorkflowTransitionEndpoint.as_view(),
        name="workflow-transitions",
    ),
    path(
        "workspaces/<str:slug>/projects/<uuid:project_id>/workflow-transitions/<uuid:pk>/",
        WorkflowTransitionEndpoint.as_view(),
        name="workflow-transitions",
    ),
    path(
        "workspaces/<str:slug>/projects/<uuid:project_id>/workflow-transitions/<uuid:workflow_transition_id>/approvers/",
        WorkflowTransitionApproverEndpoint.as_view(),
        name="workflow-transition-approver",
    ),
    path(
        "workspaces/<str:slug>/projects/<uuid:project_id>/workflow-activity/",
        WorkflowActivityEndpoint.as_view(),
        name="workflow-activity",
    ),
]

File: packages/types/src/workflow.ts
/**
 * SPDX-FileCopyrightText: 2023-present Plane Software, Inc.
 * SPDX-License-Identifier: LicenseRef-Plane-Commercial
 *
 * Licensed under the Plane Commercial License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://plane.so/legals/eula
 *
 * DO NOT remove or modify this notice.
 * NOTICE: Proprietary and confidential. Unauthorized use or distribution is prohibited.
 */

import type { TProjectBaseActivity } from "./activity";

export interface IStateTransition {
  transition_state_id: string;
  approvers: string[];
}

export interface IStateWorkFlow {
  [transitionId: string]: IStateTransition;
}

export type TStateTransitionMap = Record<string, IStateTransition>; // transitionId: IStateTransition

export interface IStateWorkFlowResponse {
  allow_issue_creation: boolean;
  transitions: TStateTransitionMap;
}

export type TWorkflowChangeHistoryFields =
  | "is_workflow_enabled"
  | "allow_work_item_creation"
  | "state_transition"
  | "state_transition_approver"
  | "reset";

export type TWorkflowChangeHistoryVerbs = "added" | "removed" | "enabled" | "disabled" | "updated";

export type TWorkflowChangeHistoryKeys = `${TWorkflowChangeHistoryFields}_${TWorkflowChangeHistoryVerbs}`;

export type TWorkflowChangeHistory = TProjectBaseActivity<TWorkflowChangeHistoryFields, TWorkflowChangeHistoryVerbs> & {
  state_id?: string | null;
  transition_state_id?: string | null;
};

export interface IStateTransitionTree {
  transition_state_ids: string[];
  approvers: string[];
}
